File: Sort/Selection_Sort/implementations/selection_sort.c
#include <stdio.h>

void selection_sort(int a[], int n);
void change(int a[], int b, int c);
void main(){
  int n;
  
  printf("n: ");
  scanf("%d", &n);

  int a[n];
  
  for(int i = 0; i < n; i++)
    scanf("%d", &a[i]);

  selection_sort(a, n);

  for(int i = 0; i < n; i++)
    printf("%d ", a[i]);
  printf("\n");
}

void selection_sort(int a[], int n){
  int smallest;
  int k;
  
  for(int i = 0; i < n - 1; i++){
    smallest = a[i];
    k = i;
    for(int j = i+1; j < n; j++)
      if(a[j] < smallest){
	smallest = a[j];
	k = j;
      }
    if(i != k)
      change(a, i, k);
  }
}

void change(int a[], int b, int c){
  int aux = a[b];
  a[b] = a[c];
  a[c] = aux;
}

File: Sort/Selection_Sort/implementations/SelectionSort.hs
module SelectionSort where

import Data.List

-- | Finds and returns the smallest element on the received list.
smallest :: (Ord a) => [a] -> a
smallest (x:[]) = x
smallest (x:y:xs)
  | x <= y = smallest (x:xs)
  | otherwise = smallest (y:xs)

-- | Sorts the list by finding the smallest element on it.
selection_sort :: (Ord a) => [a] -> [a]
selection_sort (x:[]) = x:[]
selection_sort (x:xs) = a : selection_sort (remove a (x:xs))
  where a = smallest (x:xs)

-- | Returns the same list without the element received. Doesn't removes
--   repetitions.
remove :: (Eq a) => a -> [a] -> [a]
remove _ [] = []
remove a (x:xs)
  | a == x = xs
  | otherwise = x : remove a xs
